src/plugins/dashboard/public/application/hooks/use_dashboard_app_state.ts

```diff
@@ -0,0 +1,310 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import _ from 'lodash';
+import { History } from 'history';
+import { debounceTime } from 'rxjs/operators';
+import { useCallback, useEffect, useMemo, useState } from 'react';
+import { BehaviorSubject, combineLatest, Subject } from 'rxjs';
+
+import { DashboardConstants } from '../..';
+import { ViewMode } from '../../services/embeddable';
+import { useKibana } from '../../services/kibana_react';
+import { getNewDashboardTitle } from '../../dashboard_strings';
+import { IKbnUrlStateStorage } from '../../services/kibana_utils';
+import { setDashboardState, useDashboardDispatch, useDashboardSelector } from '../state';
+import {
+  DashboardBuildContext,
+  DashboardAppServices,
+  DashboardAppState,
+  DashboardRedirect,
+  DashboardState,
+} from '../../types';
+import {
+  tryDestroyDashboardContainer,
+  syncDashboardContainerInput,
+  savedObjectToDashboardState,
+  syncDashboardIndexPatterns,
+  syncDashboardFilterState,
+  loadSavedDashboardState,
+  buildDashboardContainer,
+  loadDashboardUrlState,
+  diffDashboardState,
+  areTimeRangesEqual,
+} from '../lib';
+
+export interface UseDashboardStateProps {
+  history: History;
+  savedDashboardId?: string;
+  redirectTo: DashboardRedirect;
+  isEmbeddedExternally: boolean;
+  kbnUrlStateStorage: IKbnUrlStateStorage;
+}
+
+export const useDashboardAppState = ({
+  history,
+  redirectTo,
+  savedDashboardId,
+  kbnUrlStateStorage,
+  isEmbeddedExternally,
+}: UseDashboardStateProps) => {
+  const services = useKibana<DashboardAppServices>().services;
+  const dispatchDashboardStateChange = useDashboardDispatch();
+  const dashboardState = useDashboardSelector((state) => state.dashboardStateReducer);
+  const [lastSavedState, setLastSavedState] = useState<DashboardState>();
+
+  const [dashboardAppState, setDashboardAppState] = useState<DashboardAppState>({
```

Comment:
I'm pretty sure that the `useState` hook takes an argument as 'default' state - and the object in there is only created once when the component is mounted. I will take a bit of time to make sure of that because I definitely don't want to be creating redundant Subjects.

Reply:
Wow, looks like this was a [caveat](https://stackoverflow.com/questions/59983813/why-does-the-usestate-hook-invoke-the-initial-state-when-its-a-function-refer) with that I didn't know about. I've fixed this. Thanks for noticing this Anton, could've been bad!

Reply:
👍 
Just posting the TLDR here if anyone would bump into this conversion: 

Lazy initial state initialization: 
```
const [state, setState] = useState(() => {
  const initialState = someExpensiveComputation(props);
  return initialState;
});
```